### PR TITLE
api: Add support for subdomain allocation for Functions

### PR DESCRIPTION
**Pull request by AtilaSaraiva**
Hello folks,

This is the first commit for a series of commits adding support for functions to be allocated within a subdomain (let's call those subfunctions for brevity), and operated along other normal functions . This only first commit only adds allocation support for the Function class, I still need to add it for the TimeFunction class, not to mention that I still have to add support for it for the Operator class.
This is only but the first step in the right direction, or so I hope.

Cheers!

Authored by Átila Saraiva Quintela Soares (atilasaraiva@gmail.com)
SENAI CIMATEC geophysicist researcher.

### CODE REVIEW

File: devito/types/grid.py [hunk 1]
```diff
@@ -380,7 +383,7 @@ def __init__(self, *args, **kwargs):
             self.name = self.__class__.__name__
 
         # All other attributes get initialized upon `__subdomain_finalize__`
-        super().__init__()
+        # super().__init__()
```
Comment (FabioLuporini) [inline devito/types/grid.py:386]: leftover

**Review by FabioLuporini — COMMENTED** [review]
Thanks a lot for this! I've added it to my backlog of "PRs to be reviewed". 

At first glance, it looks good to me. Basically, you're lifting some functionality because you need the distributor machinery to be available in the SubDomains (or MultiSubDomains), right?

**Comment by FabioLuporini** [comment]
so yeah I think this is going in the right direction, my other suggestion is to keep PRs as small as possible so that we also merge them as quickly as possible :)

**Comment by AtilaSaraiva** [comment]
> so yeah I think this is going in the right direction, my other suggestion is to keep PRs as small as possible so that we also merge them as quickly as possible :)

Oh, do you mean in number of lines written, the number of commits, or number of modified files?

> Thanks a lot for this! I've added it to my backlog of "PRs to be reviewed".

No problem, thanks for the early review, but I still have to fix it for TimeFunction as well.

> At first glance, it looks good to me. Basically, you're lifting some functionality because you need the distributor machinery to be available in the SubDomains (or MultiSubDomains), right?

Yes that is correct, no big changes, just moving things around. I wonder if it would be more appropriate to split the CartesianDiscretization in to classes, one for the distributor machinery and other for the rest. Lemme know what you think.

**Comment by FabioLuporini** [comment]
> I wonder if it would be more appropriate to split the CartesianDiscretization in two classes, one for the distributor machinery and other for the rest. Lemme know what you think.

Mhh, what do you have in mind? Something like a top abstract class (CartesianDiscretization) and a mixin for "all the CartesianDiscretization that can 1) be used to define a (Time)Function and 2) be decomposed (with MPI) ?

File: devito/types/grid.py (hunk 1 above)
Comment (AtilaSaraiva) [reply devito/types/grid.py]: Sorry, fixed!

**Comment by AtilaSaraiva** [comment]
Hello Fabio! I'm studying the code for the time being, but you did say you would backport some one dimensional array functionality so I kinda need that to progress further.

Either way, I gonna need some more in depth understanding code generation and allocation to change the iteration space so Im digging the code for now.

**Comment by FabioLuporini** [comment]
> Hello Fabio! I'm studying the code for the time being, but you did say you would backport some one dimensional array functionality so I kinda need that to progress further.

ah, yes sure, in my backlog... it will take a while to get there, but we will!

> Either way, I gonna need some more in depth understanding code generation and allocation to change the iteration space so Im digging the code for now.

Sure, that makes a lot of sense to me. Get in touch on slack if you have any questions (feel free to use the #development channel)

**Comment by FabioLuporini** [comment]
what's the state of this?

**Comment by FabioLuporini** [comment]
no followup, closing

**Pull request closed without merging**